designs/odcr.md

```diff
@@ -0,0 +1,140 @@
+checking how close to limit of the reservation.
+
+```yaml
+Name:               example-node
+Labels:             karpenter.k8s.aws/capacity-reservation-id=cr-12345
+                    karpenter.sh/capacity-type=on-demand
+```
+
+`karpenter.k8s.aws/capacity-reservation-id` will be the capacity reservation the node launched from. 
+
+We will propagate this information via [instance](https://github.com/aws/karpenter-provider-aws/blob/main/pkg/providers/instance/types.go#L29) by extracting it from [DescribeInstance](https://github.com/aws/karpenter-provider-aws/blob/main/pkg/batcher/describeinstances.go#L48) [aws doc]([https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#EC2.DescribeInstances](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_CapacityReservationSpecificationResponse.html)https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_CapacityReservationSpecificationResponse.html).
+
+### Failed to launch Nodes into Capacity Reservation
+
+The main failure scenario is when Capacity Reservation limit is hit and no new nodes can be launched from any Capacity Reservation the launch template targets. 
+
+#### Status
+
+We will expose this information both in the NodeClaim's, EC2NodeClass' and NodePool's Condition Status. _I know that currently NodeClasses and NodePool status do not have conditions but I wanted to see if we are opened to adding Conditions to these resources_
+
+```yaml
+Status:
+  Conditions:
+    Last Transition Time:  2024-02-22T03:46:42Z
+    Status:                LimitExceeded
+    Type:                  CapacityReservation
+```
+The condition will reset if new nodes were able to launch and the Status will return to `Available`. It may also be helpful to show the
+utilization of the capacity if we are interested in perform consolidation into Capacity Reserved instances
+
+#### Error handling
+
+We will avoid updating (unavailableOfferingCache)[https://github.com/aws/karpenter-provider-aws/blob/main/pkg/providers/instance/instance.go#L239C41-L239C58] because the pool is different than rest of AWS. However we may want to create a new unavailable offering cache keyed against Capacity Reservations. _Not sure if we want to support to this during the first iteration_ 
+
+## Open Questions
+- The UX of adding Capacity Reservation feels wrong because NodeClasses previously didn't fully restrict instance types but with Capacity Reservation it kind of does. There isn't a good primative in Karpenter to expose these kinds of restrictions (I think?) specifically around preflight or static analysis that tells you your NodePool may not be able to launch any nodes because of X, Y and Z in your EC2NodeClass. I believe this is already an issue where if a NodeClass selects for x86 architecture AMIs but the NodePool allows for ARM architecture instance types that Karpenter may just quietly never spawn ARM instances?
```

Comment:
There's a primitive. When you are resolving the GetInstanceTypes() call, you can just tell it not to return back certain instance types based on the EC2NodeClass used. We do have some slight mutation behavior around some EC2NodeClasses specified today already, like when using Windows, this scopes down the supported instance types that we return back.